Expand `GeantPhysicsOptions` to encompass all physics processes

stognini opened the pull request:
This PR expands `GeantPhysicsOptions` to allow the selection of every physics process currently available. The `simple-driver.py` generates the json accordingly and the ROOT output file stores all `RunnerInput` variables to make the validation process safer.

stognini committed:
Expand Geant physics options

stognini committed:
Update simple-driver

stognini committed:
Expand ROOT output; Fix Runner bug

stognini committed:
Update physics list documentation

sethrj commented on app/demo-loop/RootOutput.cc:
```diff
@@ -44,14 +44,39 @@ void write_to_root(RunnerInput const& cargs,
+    tree_input->Branch("msc_safety_factor", &phys.msc_safety_factor);
+    tree_input->Branch("msc_lambda_limit", &phys.msc_lambda_limit);
+    tree_input->Branch("apply_cuts", &phys.apply_cuts);
+
```
I'm doubting whether it's the right approach to maintain this ROOT output parameter list in addition to the JSON output. I agree that it's good to have a copy of the input for provenance, but how about instead make a single "output" branch, and use the `OutputRegistry` to write a single string with the JSON output at the end of the simulation? That way we can rely on the existing infrastructure, get more data out, and remove all this code.

stognini replied:
We could store a single, large string. The downside is that it makes it harder to look up stuff down the road. For example, we can't just ask root something like: `tree->GetLeaf("brems")->GetValue()`. Can we create a json reader from a single string, so we have an interface? If that's trivial, the single string in the root file seems like a good option.

sethrj replied:
Yeah, if you're using ROOT you should still be able to use C++ libraries like NLJSON, right?  (Especially if they're header-only.) In that case you should be able to do 
```c++
auto output = nlohmann:json::parse(tree->GetLeaf("output")->GetValue();
output["input"]["brems"];
```
and bonus, you'll also be able to check the names of actions, particles,  processes, etc.

sethrj replied:
Actually I think ROOT *requires* nljson these days, so it should definitely be available... maybe @pcanal can comment.

stognini replied:
Great. That works nicely.

stognini replied:
> Actually I think ROOT requires nljson these days

I don't know if it is a _requirement_, but it has nljson (either externally, or built in), so my only worry was parsing a string instead of an ifstream...

sethrj commented on app/demo-loop/Runner.cc:
```diff
@@ -77,17 +77,19 @@ Runner::Runner(RunnerInput const& inp,
+        // Initialize ROOT file and step collector
+        root_manager_
+            = std::make_shared<RootFileManager>(inp.mctruth_filename.c_str());
+        this->build_step_collectors(inp);
```
No. We need to call this even if ROOT isn't active: there are other step collectors. I don't understand why this code was changed here and below?

stognini replied:
Ohh good point. The refactoring got me a bit confused.
But I moved the construction of `root_manager_` here to separate itfrom the step collector function, so it was more readable: construct root file manager, build step collectors, write input/params data to root -- the last bit was not working (we were saving only the `steps` tree to the root file).

sethrj replied:
Oh I see the problem looking at `Runner::build_step_collectors`: I accidentally declared local variables `root_manager_` and `step_collector_` that *shadowed* the class data.

I think it's OK to move the root manager construction into the main constructor (and make sure to delete my bad local variables!), but leave everything else the way it was.

sethrj commented on src/celeritas/ext/GeantPhysicsOptions.hh:
```diff
@@ -76,27 +80,40 @@ enum class RelaxationSelection
     real_type msc_lambda_limit{0.1};  // 1 mm
-    bool apply_cuts{false};
 
+    bool apply_cuts{false};
```
Maybe move this parameter next to `lowest_electron_energy` since it's related to that?

sethrj commented on src/celeritas/ext/detail/GeantPhysicsList.cc:
```diff
@@ -215,37 +215,38 @@ void GeantPhysicsList::add_gamma_processes()
     auto* physics_list = G4PhysicsListHelper::GetPhysicsListHelper();
 
-    if (p == G4Positron::Positron())
+    if (options_.annihilation)
```
```suggestion
    if (options_.annihilation && p == G4Positron::Positron())
```
and remove the extra braces.

sethrj commented on src/celeritas/ext/detail/GeantPhysicsList.cc:
```diff
@@ -215,37 +215,38 @@ void GeantPhysicsList::add_gamma_processes()
     }
 
-    if (true)
+    if (options_.ionization)
```
If ionization is off, do we still correctly get dE/dx from brems? Or is our along-step data zeroed out?

stognini replied:
I believe it's zeroed out. @whokion can you confirm?

sethrj replied:
Ping @whokion : can we correctly get radiative slowing-down data when disabling ionization?

amandalund replied:
If I recall correctly, with ionization disabled we'll still get the (brems contribution to) dE/dx from the brems process: in that case `IsIonisationProcess()` will be true for brems.

sethrj commented on src/celeritas/ext/GeantPhysicsOptions.hh:
```diff
@@ -76,27 +80,40 @@ enum class RelaxationSelection
+    bool annihilation{true};
+    bool brems{true};
+    BremsModelSelection brems_selection{BremsModelSelection::all};
+    MscModelSelection msc{MscModelSelection::urban};
```
This default changed and I don't think it should have.

sethrj commented on src/celeritas/ext/GeantPhysicsOptions.hh:
```diff
@@ -76,27 +80,40 @@ enum class RelaxationSelection
+    // Electrons and positrons
+    bool ionization{true};
+    bool annihilation{true};
+    bool brems{true};
+    BremsModelSelection brems_selection{BremsModelSelection::all};
```
These should be rolled into a single `brems` parameter: add a new `none` value.

sethrj commented on src/celeritas/ext/GeantPhysicsOptions.hh:
```diff
@@ -76,27 +80,40 @@ enum class RelaxationSelection
+    // Gamma
     bool coulomb_scattering{false};
+    bool compton_scattering{true};
+    bool photoelectric{true};
```
The new parameters need to be added to `GeantPhysicsOptions.json.cc`.

stognini replied:
Good catch. They are in the null process branch. Moving them in...

stognini committed:
Address comments

stognini committed:
Revert to single input tree

sethrj commented on src/celeritas/ext/GeantPhysicsOptions.hh:
```diff
@@ -76,26 +81,38 @@ enum class RelaxationSelection
+    // Electrons and positrons
+    bool ionization{true};
+    bool annihilation{true};
+    BremsModelSelection brems_selection{BremsModelSelection::all};
```
Restore as `brems`

sethrj commented on src/celeritas/ext/GeantPhysicsOptions.hh:
```diff
@@ -105,20 +122,33 @@ struct GeantPhysicsOptions
            && a.msc_safety_factor == b.msc_safety_factor
            && a.msc_lambda_limit == b.msc_lambda_limit
-           && a.apply_cuts == b.apply_cuts && a.verbose == b.verbose;
+           && a.verbose == b.verbose;
```
Verbose was intentionally omitted because it doesn't change the loaded physics. I should've added a comment to that effect

stognini replied:
I think verbose was always in there, I just moved it to a new line for readability. But I agree that for an equality operator we could drop that.

sethrj replied:
Oh. I didn't notice 😅 Thanks.

sethrj commented on app/demo-loop/RootOutput.cc:
```diff
@@ -29,33 +31,21 @@ void write_to_root(RunnerInput const& cargs,
-    tree_input->Branch("sync", &args.sync);
-    tree_input->Branch("step_limiter", &args.step_limiter);
+    // Input
+    nlohmann::json json_inp;
+    to_json(json_inp, cargs);
+    std::string str_input(json_inp.dump());
```
I think this should work:
```suggestion
    std::string str_input(nlohmann::json(cargs).dump());
```

sethrj commented on app/demo-loop/RootOutput.cc:
```diff
@@ -29,33 +31,21 @@ void write_to_root(RunnerInput const& cargs,
+    auto const& gpo = cargs.geant_options;
+    nlohmann::json json_phys;
+    to_json(json_phys, gpo);
+    std::string str_phys(json_phys.dump());
```
Same as above. Also, you're only interested in manually storing the input arguments, as opposed to storing all the Celeritas JSON output data?

sethrj commented on app/demo-loop/Runner.cc:
```diff
@@ -77,17 +77,18 @@ Runner::Runner(RunnerInput const& inp,
-        // Store input and CoreParams data
+        // Initialize ROOT file; Store input and core params
+        root_manager_
+            = std::make_shared<RootFileManager>(inp.mctruth_filename.c_str());
```
I'm thinking about this again... if we're *only* going to create a ROOT manager if we're making an MC truth file, then this should go in the `build_step_collectors`. I don't like having the big block here required to be among the `this->build` list... maybe make a `this->build_root` method...

stognini committed:
Address comments

- Move RootFileManager construction to Runner::build_step_collectors
- Refactor brems_selection -> brems
- Simplify RootOutput::write_to_root

stognini committed:
Fix test

stognini committed:
Merge branch 'develop' into phys-opt

sethrj reviewed (approved):
Nice! Good step toward #670 .

sethrj commented:
Hmmm, why are these tests failing: https://cloud.cees.ornl.gov/jenkins-ci/blue/organizations/jenkins/Celeritas/detail/PR-744/8/tests

Looks like ROOT is segfaulting as the app quits?

stognini commented:
>Looks like ROOT is segfaulting as the app quits?

Somehow the `ActionDiagnostics` is borking the step collector?

- If I do not run `build_diagnostics` in the `Runner` constructor, everything works fine.
- If I keep the diagnostics, but do not call any of the `write_to_root()`, the `demo-loop` finishes correctly, but the ROOT file does not save the `steps` tree (although its size is about right).
- If any of the `write_to_root()` is called, we either get a `segfault` or a `malloc: Incorrect checksum for freed object 0xNNN: probably modified after being freed` error, depending on which is on.

stognini committed:
Add workaround to fix segfault

stognini committed:
Merge branch 'develop' into phys-opt

stognini committed:
Clean up

sethrj commented on app/demo-loop/Runner.cc:
```diff
@@ -361,13 +353,15 @@ void Runner::build_primaries(RunnerInput const& inp)
         root_manager_
             = std::make_shared<RootFileManager>(inp.mctruth_filename.c_str());
+        write_to_root(inp, root_manager_.get());
+        // Must be called after all actions were added to action_registry
```
This is why I put the ROOT writing at the end of the constructor. Can you move it back there? I think deleting the stray `root_manager_` local function will fix this all.

sethrj commented on app/demo-loop/Runner.hh:
```diff
@@ -101,7 +101,6 @@ class Runner
 
   private:
     //// DATA ////
-
```
I kind of prefer blank lines after these grouping separators: is this whitespace change consistent with what's in the rest of the code?

stognini replied:
Mostly yes. I can refactor and add spaces between `//// [type] ////` and the code in any missing class...

sethrj replied:
I don't think you need to refactor; this just looks like a standalone change (probalby introduced by adding something and deleting it) that is unnecessary.

stognini replied:
Yes. My point is that there are classes (ProcessBuilder, GeantModelImporter, and others) with no blank lines. So I could make all of them consistent now.

sethrj replied:
No, let's do it later if we have a separate formatting commit. I'm fine with minor inconsistencies, I don't like having one file with a one-line whitespace change when nothing else in the file is modified.

sethrj commented on src/celeritas/ext/RootFileManager.cc:
```diff
@@ -59,13 +59,15 @@ RootFileManager::make_tree(char const* name, char const* title)
     int write_status = tfile_->Write();
     CELER_ENSURE(write_status);
+    tfile_->Close();
+    CELER_ENSURE(!tfile_->IsOpen());
```
I don't think we want this. The ROOT file should be open as long as the RFM exists: we don't want to have a separate "exists and works" and "exists and doesn't work" state. This is one of the fundamantal tenets of good c++ programming: classes should always be in a valid state. And I don't think this class needs the extra complexity of e.g. `ofstream` where you have the ability to close before destruction, separate accessors for validity, etc.

stognini replied:
Ohh thanks! Good catch. Yes, that was there just for testing and I missed reverting this.

stognini committed:
Address comments

stognini committed:
Remove unnecessary includes

sethrj reviewed (approved):
Thanks for your help debugging!

stognini commented:
And thank you for figuring out the pointer cycle issue and fixing it.
